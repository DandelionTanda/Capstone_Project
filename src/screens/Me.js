import { useState, useEffect }  from "react";
import * as React from 'react';
import { ActivityIndicator, Button, Pressable, FlatList, Dimensions, SafeAreaView, StatusBar, StyleSheet, Text, TouchableOpacity,View,Icon,Image, ScrollView } from "react-native";
import 'react-native-gesture-handler';
import { NavigationContainer, getFocusedRouteNameFromRoute, } from '@react-navigation/native';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { createStackNavigator } from '@react-navigation/stack';
import { MaterialCommunityIcons, Ionicons } from '@expo/vector-icons';
import { fetchUser } from '../networking/Api'

export default function Me( { navigation } ) {
  const [user, setUser] = useState(null)
  const [loading, setLoading] = useState(true)
  const [error, setError] = useState('') 
  function clear(){
    navigation.navigate('Login')
    localStorage.clear()
  }

  useEffect(async () => {  
    let user = await fetchUser()    
    await setUser(user)  
    await setLoading(false)        
  },[])

  const onRefresh = React.useCallback(async () => {
    await setLoading(true);
    let user = await fetchUser()   
    if (user instanceof Error) {
      setError("ooops, there is an error from server")
    }
    else {
      setError("")
      await setUser(user)  
    }
    await setLoading(false)
  }, []);  
  
  if (loading !== true){  
    if (error) {
      return (
        <View style={{ flex: 1,justifyContent: 'center', alignItems: 'center',
        flexDirection: 'column',}}>
          <Text style={styles.error}>{error.message}</Text>
          <Pressable style={styles.button} onPress={onRefresh}>
            <Text style={{color:'white', fontSize:20}}>Refresh</Text>
          </Pressable>
        </View>
      ) 
    } else {
      return (
        <ScrollView style={{ flex: 1}}>
          {/* View for image */}
          <View style={styles.image}>
          {user.photo!=='null'?
          <Image style={{ width: 150, height: 150, borderRadius: 130/2 }} 
          source={{
              uri: user.photo,
          }} alt = "Avatar"></Image>:
          <MaterialCommunityIcons name="account-circle"  size={100}/>}
          </View>
          {/* View for information: Name, Email, ID, Role, Company*/}
          <View style= {styles.personalInfor}>
            <Text style={styles.label}>Name</Text>
            <Text style={styles.infor}>{user.name}</Text>
          </View>
          <View style= {styles.personalInfor}>
            <Text style={styles.label}>Email</Text>
            <Text style={styles.infor}>{user.email}</Text>
          </View>
          <View style= {styles.personalInfor}>
            <Text style={styles.label}>Company</Text>
            <Text style={styles.infor}>{user.organisation}</Text>
          </View>
          <View style= {styles.personalInfor}>
            <Text style={styles.label}>Employee ID</Text>
            <Text style={styles.infor}>{user.id}</Text>
          </View>
          {/* logout button */}
          <TouchableOpacity
            style={styles.logoutButton}
            onPress={clear}
          >
            <Ionicons name="log-out" size={35} color="white"/>
            <Text style={{fontSize: 20, fontWeight: "bold", color: 'white'}}>Log out</Text>
          </TouchableOpacity>
        </ScrollView>
      )
    }
  } 
  else {
      return (
        <View style={[styles.container, styles.horizontal]}>      
          <ActivityIndicator size="large" color="#0000ff"/>
        </View>
        )
  }
    
}

// Retrieve initial screen's width
let width = Dimensions.get('screen').width;

// Retrieve initial screen's height
let height = Dimensions.get('screen').height;

{/*Custom styles*/}
let styles;
if(width < height){
  //small screen
  if(width<350){
    styles = StyleSheet.create({
      container: {
        flex: 1,    
        backgroundColor: '#F5F3F3',
        marginTop: 24,
        marginLeft: 24,
        marginRight: 24,
        marginBottom: 24
      },
      horizontal: {
        flexDirection: "row",
        justifyContent: "space-around",
        padding: 10
      },
      logoutButton: {
        alignSelf:"center",
        marginTop: 24,
        alignItems: "center",
        backgroundColor: "#E3310E",
        padding: 10,   
        flexDirection: 'row',
        borderTopLeftRadius: 15,
        borderTopRightRadius: 15,
        borderBottomLeftRadius: 15,
        borderBottomRightRadius: 15,
        marginBottom:28
      },
      personalInfor:{
        marginLeft:30,
        marginRight:30,
        borderBottomColor: 'black',
        borderBottomWidth: 1
      },
      label:{
        marginTop:10,
        fontSize: 20,
        fontWeight: "bold",
        color: '#1D87E3'
      },
      infor:{
        fontSize: 16,
        color: '#232A22',
        lineHeight: 32
      },
      image:{
        flexDirection: "row",
        marginTop:28 ,
        alignSelf:'center',
      },error: {
        fontSize:20,
        alignItems: "center",
        justifyContent: 'center',
        color:'red',
      },
      button:{
        alignItems: 'center',
        justifyContent: 'center',
        marginLeft:20,
        marginRight:20,
        paddingVertical: 16,
        borderRadius: 4,  
        backgroundColor: '#45B8DB',
        position:"relative",
        top:20,
        width: '40%'
      },
    });
  }
  //large screen
  else{
    styles = StyleSheet.create({
      container: {
        flex: 1,    
        backgroundColor: '#F5F3F3',
        marginTop: 24,
        marginLeft: 24,
        marginRight: 24,
        marginBottom: 24
      },
      horizontal: {
        flexDirection: "row",
        justifyContent: "space-around",
        padding: 10
      },
      logoutButton: {
        alignSelf:"center",
        marginTop: 30,
        alignItems: "center",
        backgroundColor: "#E3310E",
        padding: 10,   
        flexDirection: 'row',
        borderTopLeftRadius: 15,
        borderTopRightRadius: 15,
        borderBottomLeftRadius: 15,
        borderBottomRightRadius: 15,
        marginBottom:40
      },
      personalInfor:{
        marginLeft:30,
        marginRight:30,
        borderBottomColor: 'black',
        borderBottomWidth: 1
      },
      label:{
        marginTop:10,
        fontSize: 22,
        fontWeight: "bold",
        color: '#1D87E3'
      },
      infor:{
        fontSize: 18,
        color: '#232A22',
        lineHeight: 40
      },
      image:{
        flexDirection: "row",
        marginTop:40 ,
        alignSelf:'center',
      },error: {
        fontSize:20,
        alignItems: "center",
        justifyContent: 'center',
        color:'red',
      },
      button:{
        alignItems: 'center',
        justifyContent: 'center',
        marginLeft:20,
        marginRight:20,
        paddingVertical: 16,
        borderRadius: 4,  
        backgroundColor: '#45B8DB',
        position:"relative",
        top:20,
        width: '40%'
      },
    });
  }
}