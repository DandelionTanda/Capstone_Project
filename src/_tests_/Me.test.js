import 'react-native'
import React from 'react'
import Me from "../screens/Me"

import renderer from 'react-test-renderer';
import {fireEvent, render, waitFor} from '@testing-library/react-native'
import {fakeUser} from './fakeData'

describe('Login setup', () => {
  beforeEach(() => {   
    global.fetch = jest
    .fn()
    .mockImplementation(() => Promise.resolve(
      Promise.resolve({
        ok: true,
        status: 200,       
        json: () => Promise.resolve(fakeUser)
      })
    )); 
        
  });
  it('Should render correctly', async () => {
    const { toJSON } = render(
      <Me />
    );
    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    })
  })

  it('Should display user information', async () => {
    const { queryByText } = render(
      <Me />
    );
    await waitFor(() => {
      expect(queryByText("Weiran Zou")).toBeTruthy();
      expect(queryByText("zouweiran9122@gmail.com")).toBeTruthy();
      expect(queryByText("QUT Capstone - DEMO ACCOUNT")).toBeTruthy();
      expect(queryByText("1748964")).toBeTruthy();
    })
  })

  it('should render a logout button', async () => {
    const { getByText } = render(
      <Me/>
    );
    await waitFor(() => {
      expect(getByText('Log out')).toBeTruthy();
    })
  });
  /*
  it('on pressing Logout button, a press handler function should be triggered on click event',  () => {
    const onPressMock = jest.fn();
    const { getByText } = render(
      <Me/>
    );
    fireEvent.press(getByText('Log out'));
    expect(onPressMock).toHaveBeenCalled();
  });  
  */
  it('Expect to clear token in localStorage & navigate to login screen after clicking on log out button', async () => {
    // Mocking navigate method
    const navigate = jest.fn();
   
    const { getByText } = render(
      <Me navigation={{ navigate }}/>
    );
    await waitFor(() => {
      fireEvent.press(getByText('Log out'));
    })
    expect(localStorage.clear).toHaveBeenCalledTimes(1);
   // Asserting screen navigation
    expect(navigate).toHaveBeenCalledTimes(1);
    expect(navigate).toHaveBeenCalledWith('Login');
  });  
})