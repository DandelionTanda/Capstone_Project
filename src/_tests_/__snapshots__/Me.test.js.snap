// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login setup Should render correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "flexDirection": "row",
          "marginTop": 40,
        }
      }
    >
      <Image
        alt="Avatar"
        source={
          Object {
            "uri": "https://s3.amazonaws.com/PayAus/logins/photos/214/946/904/icon/bd8b9cf053c843329b270cd902dd1cb8.jpeg",
          }
        }
        style={
          Object {
            "borderRadius": 65,
            "height": 150,
            "width": 150,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "marginLeft": 30,
          "marginRight": 30,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1D87E3",
            "fontSize": 22,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Name
      </Text>
      <Text
        style={
          Object {
            "color": "#232A22",
            "fontSize": 18,
            "lineHeight": 40,
          }
        }
      >
        Weiran Zou
      </Text>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "marginLeft": 30,
          "marginRight": 30,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1D87E3",
            "fontSize": 22,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Email
      </Text>
      <Text
        style={
          Object {
            "color": "#232A22",
            "fontSize": 18,
            "lineHeight": 40,
          }
        }
      >
        zouweiran9122@gmail.com
      </Text>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "marginLeft": 30,
          "marginRight": 30,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1D87E3",
            "fontSize": 22,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Company
      </Text>
      <Text
        style={
          Object {
            "color": "#232A22",
            "fontSize": 18,
            "lineHeight": 40,
          }
        }
      >
        QUT Capstone - DEMO ACCOUNT
      </Text>
    </View>
    <View
      style={
        Object {
          "borderBottomColor": "black",
          "borderBottomWidth": 1,
          "marginLeft": 30,
          "marginRight": 30,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#1D87E3",
            "fontSize": 22,
            "fontWeight": "bold",
            "marginTop": 10,
          }
        }
      >
        Employee ID
      </Text>
      <Text
        style={
          Object {
            "color": "#232A22",
            "fontSize": 18,
            "lineHeight": 40,
          }
        }
      >
        1748964
      </Text>
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#E3310E",
          "borderBottomLeftRadius": 15,
          "borderBottomRightRadius": 15,
          "borderTopLeftRadius": 15,
          "borderTopRightRadius": 15,
          "flexDirection": "row",
          "marginBottom": 40,
          "marginTop": 30,
          "opacity": 1,
          "padding": 10,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": "white",
              "fontSize": 35,
            },
            undefined,
            Object {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        ïŽŽ
      </Text>
      <Text
        style={
          Object {
            "color": "white",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        Log out
      </Text>
    </View>
  </View>
</RCTScrollView>
`;
